### ISSUE TITLE
LKMM dependency tracking issues

### ISSUE THREAD

**Issue by db7** [MPI-SWS/genmc#23]
Here is small example that seems to identify issues in the LKMM dependency tracking of GenMC.

```c
#include <pthread.h>
#include <assert.h>
#include <stdbool.h>
#include <lkmm.h>
void __VERIFIER_spin_start(void);
void __VERIFIER_spin_end(bool);
#define await_while(cond)                                        \
	for (int tmp = 0; __VERIFIER_spin_start(), tmp = (cond), \
			__VERIFIER_spin_end(!tmp), tmp;)
/******************************************************************************/
int locked;
int *next;
void *alice(void *arg)
{
	WRITE_ONCE(locked, 1);
	smp_store_release(&next, &locked);
	await_while (READ_ONCE(locked));
	return 0;
}
void *bob(void * arg)
{
	int *i;
	await_while((i = READ_ONCE(next)) == 0);
	WRITE_ONCE(*i, 0);
	return 0;
}
/******************************************************************************/
int main()
{
	pthread_t ta, tb;
	pthread_create(&ta, 0, alice, 0);
	pthread_create(&tb, 0, bob, 0);
	pthread_join(ta, 0);
	pthread_join(tb, 0);
	assert(locked == 0);
	return 0;
}

```

GenMC detects the following liveness violation:

```
user@d0d1f66d6eaf:~$ genmc -check-liveness -disable-spin-assume -disable-load-annotation -lkmm test-mp.c 
WARNING: LKMM support is still at an experimental stage!
Error detected: Liveness violation!
Event (2, 8) in graph:
<-1, 0> main:
	(0, 0): B
	(0, 1): TC [forks 1] L.40
	(0, 2): TC [forks 2] L.41
	(0, 3): TJ L.42
<0, 1> alice:
	(1, 0): B
	(1, 1): M
	(1, 2): Wrlx (locked, 1) L.19
	(1, 3): Wna (, 0x39720e0) L.20
	(1, 4): Rna (, 60236000) [(1, 3)] L.20
	(1, 5): Wrel (next, 60236000) L.20
	(1, 6): SPIN_START L.21
	(1, 7): Rrlx (locked, 1) [(1, 2)] L.21
<0, 2> bob:
	(2, 0): B
	(2, 1): M
	(2, 2): SPIN_START L.29
	(2, 3): Rrlx (next, 60236000) [(1, 5)] L.29
	(2, 4): Wna (, 60236000) L.29
	(2, 5): Rna (, 0x39720e0) [(2, 4)] L.29
	(2, 6): Wrlx (locked, 0) L.31
	(2, 7): D L.32
	(2, 8): E

Non-terminating spinloop
Number of complete executions explored: 0
Number of blocked executions seen: 1
Total wall-clock time: 0.02s
```

However, Alice should **not** hang. As I understand, (2,6) must happen after (1,2) because
1. (1,2) happens before (1,5)
2. (1,5) is release
3. (2,3) reads from (1,5)
4. there is an address dependency between (2,3) and (2,6)

Interestingly, if I run the same example with the flags `-check-consistency-point=step` and `-check-consistency-type=full`, then GenMC does not detect the liveness violation:
```
genmc -check-liveness -disable-spin-assume -disable-load-annotation -lkmm -check-consistency-point=step -check-consistency-type=full test-mp.c 
WARNING: LKMM support is still at an experimental stage!
Number of complete executions explored: 1
Number of blocked executions seen: 2
Total wall-clock time: 0.03s
```

Also, if I use the automatic spin-assume transformation (*instead of await_while+disable flags*), the presumably incorrect liveness violation is still detected even with the `-check-consistency` flags above.

Is my understanding of the address dependency/LKMM incorrect or is this an issue with the dependency tracking in GenMC? 

Which flags should I use for being sure I won't have false positives?

**Comment by JonasOberhauser** [comment]
This is probably a bug in the liveness checking. I assume that a consistency check before reporting the liveness violation has been forgotten.

**Comment by michaliskok** [comment]
There are two different bugs here, and both are neither LKMM-specific nor related to dependency tracking, AFAICT.

1. Manual instrumentation: the issue is the (implicit) usage of wb. Both stores to `locked` are wb-maximal, and GenMC (erroneously) reports the execution above as a liveness violation. I am not sure whether this is worth fixing; it could unnecessarily slow down the tool when liveness checks are enabled. I am more tempted to disable liveness checks under wb, although I am not sure yet.

GenMC should not report a liveness violation when `-mo` is used for the manual instrumentation.

2. Spin-assume: this is because GenMC takes the wrong events into account when checking for liveness. I will push a fix as soon as possible. (It might still take some days as I am currently out of office.)

**Comment by db7** [comment]
> There are two different bugs here, and both are neither LKMM-specific nor related to dependency tracking, AFAICT.
> 
> 1. Manual instrumentation: the issue is the (implicit) usage of wb. Both stores to `locked` are wb-maximal, and GenMC (erroneously) reports the execution above as a liveness violation. I am not sure whether this is worth fixing; it could unnecessarily slow down the tool when liveness checks are enabled. I am more tempted to disable liveness checks under wb, although I am not sure yet.
> 
> GenMC should not report a liveness violation when `-mo` is used for the manual instrumentation.

Yes, I confirm that. Thank you. I incorrectly assumed `-mo` was automatically enabled when using `-check-liveness`. If `-check-liveness` require `-mo`, I think GenMC should either exit with an error when `-mo` is not given or should automatically enable it.

> 1. Spin-assume: this is because GenMC takes the wrong events into account when checking for liveness. I will push a fix as soon as possible. (It might still take some days as I am currently out of office.)

That is ok from our side, we are not using the spin-assume transformation right now.

Thanks for your reply.

**Comment by michaliskok** [comment]
Fixed in v0.6.1. (Liveness checks are now only allowed under `-mo`.)
